### PR TITLE
iOS Text Editing Infinite Loop

### PR DESCRIPTION
**Pull request by justinmc**
## Description

Programmatically modifying text while it is being input causes an infinite loop on iOS, but not on Android.

For example, this is reproduced when setting an onChanged handler and adding a character to the text.  See the linked issue for more details.

The reason why Android doesn't cause an infinite loop is that Android batches text changes that come very quickly.  This PR solves the problem on iOS by ~~performing a similar batching by debouncing.~~ eliminating the ACK response that the engine was sending back to the framework.

## Related Issues

Closes https://github.com/flutter/flutter/issues/61282

Further work in this area:
https://github.com/flutter/engine/pull/21534
https://github.com/flutter/flutter/pull/66864

## Tests

I added the following tests:

  * Updated tests that depended on the ACK response.

## Breaking Change

Not a breaking change.

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
Is possible to fix this by holding off the framework -> engine `TextEditingValue` update until `onChange` is called? Or there're other manifestations of this bug (other than changing the text editing value in `onChange`)?

**Comment by justinmc** [comment]
@LongCatIsLooong Good thinking. I just tried playing around with that idea and it didn't seem to work, though (messing with the calls around [here](https://github.com/flutter/flutter/blob/fb08acf10c63c6eaeb6f32931a965f18cbf28eb9/packages/flutter/lib/src/widgets/editable_text.dart#L2145)).

The reason it doesn't work seems to be that the problem is more with the engine sending updates to the framework too fast.  Here's an example:

  * Framework receives "a" from the engine, changes it to "a0", and sends it back to the engine.
  * Framework receives "ab" from the engine before the "a0" ack because the typing was so fast. It changes it to "ab0" and sends that to the engine.
  * Framework receives the engine's ack of its first message, "a0". Comparing that to its latest value, "ab0", it concludes that it needs to call onChanged and another zero is added and sent to the engine: "a00".
  * Framework receives the engine's ack of its second message, "ab0", calls onChanged and sends the engine "ab00".
  * Infinite loop where messages alternate between "a000..." and "ab000..." with increasing zeroes.

Maybe this could be fixed in the framework with a similar sort of batching that is being done in the engine in this PR, or a more sophisticated queue that waits for ack responses from the engine?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 1]
```diff
@@ -420,6 +420,7 @@ @implementation FlutterTextInputView {
   int _textInputClient;
   const char* _selectionAffinity;
   FlutterTextRange* _selectedTextRange;
+  NSDictionary* _latestState;
```
[the remark this answers is not in the document]
Comment (justinmc) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:423]: Alright got the deallocation in 9949623.  For erasing the state each time the connection changes, would that just be in [resetAllClientIds](https://github.com/flutter/engine/blob/aec855d3963c9cf0e7082b4d05288c75b4bc36e1/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L1319)?

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm:423]: `FlutterTextInputView.setClientId` I think?

**Comment by LongCatIsLooong** [comment]
> Framework receives the engine's ack of its first message, "a0". Comparing that to its latest value, "ab0", it concludes that it needs to call onChanged and another zero is added and sent to the engine: "a00".

I think we should probably revisit this approach (sending the ACK).  Context: https://github.com/flutter/engine/pull/3995. If the engine does not ACK, or the framework knows it's just an ACK message, I think we can prevent the feedback loop too right?

**Comment by justinmc** [comment]
Yeah good point, looking at that old PR it seems like this setup is less set in stone than I thought.  I also tried this again on Mac and I was able to reproduce the infinite loop there, too.  It wouldn't be ideal to have to create this debouncing approach in multiple different platforms.

I'll try to see what it would take to clean up this ack message system itself.

**Comment by justinmc** [comment]
@LongCatIsLooong Alright the latest code is my experiment in just not sending the ACK.  It seems to prevent the infinite loop, and I can't find anything else that it breaks.  What do you think about this approach?

**Comment by LongCatIsLooong** [comment]
Sorry didn't see the comment. Taking a look.

**Review by LongCatIsLooong — APPROVED** [review]
LGTM. Thank you! 

This line in the test file may need updating too:
https://github.com/flutter/engine/blob/255efbbf18d844ee953d6c66e82127901096908a/shell/platform/darwin/ios/framework/Source/FlutterTextInputPluginTest.m#L19

**Pull request merged**